### ISSUE TITLE
[Data] Potential repetitive read + data processing + sort computation as part of pipeline that runs multiple sorts

### ISSUE THREAD

**Issue by afang-story** [ray-project/ray#39753]
### What happened + What you expected to happen

Trying to run multiple push-based sorts (through groupby) in a row, however it does the initial data reading + preprocessing  + sort once, before doing the entire pipeline a separate time (now including the second sort and what comes after).

Potentially related to LimitOperator, operation fusion, or just in general running two sorts in a single pipeline.

Logs include the following:

Executing DAG InputDataBuffer[Input] -> TaskPoolMapOperator[ReadBinary -> MapBatches(tar_to_entries)] -> LimitOperator[limit=1].  - this takes no time and seems to be skipped

Executing DAG InputDataBuffer[Input] -> TaskPoolMapOperator[ReadBinary -> MapBatches(tar_to_entries)] -> AllToAllOperator[Sort] -> TaskPoolMapOperator[MapBatches(group_fn)] -> LimitOperator[limit=1] - this actually takes time and gets executed

Executing DAG InputDataBuffer[Input] -> TaskPoolMapOperator[ReadBinary -> MapBatches(tar_to_entries)] -> AllToAllOperator[Sort] -> TaskPoolMapOperator[MapBatches(group_fn)] -> AllToAllOperator[Sort] -> TaskPoolMapOperator[MapBatches(group_fn)] - this is the complete pipeline and gets executed separately from the above line

### Versions / Dependencies

pip install -U "ray[default] @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-3.0.0.dev0-cp310-cp310-manylinux2014_x86_64.whl"
pip install boto3==1.26.90
pip install s3fs==2022.11.0
pip install psutil
pip install pysimdjson
pip install pandas
pip install pyarrow

### Reproduction script

Longer, so sent separately to @stephanie-wang 

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by stephanie-wang** [comment]
It looks like the issue here is that when we have multiple sort passes (e.g., groupby -> map_groups -> groupby), the sort in the second groupby ends up re-executing the first groupby. Not sure yet if this is because it's trying to get the schema, or if it has to do with the sort sampling stage.

It can be easily fixed by putting a .materialize() before the second groupby (groupby -> map_groups -> materialize -> groupby)

**Comment by c21** [comment]
I think it's probably related to the `schema()` call inside `groupby` - https://github.com/ray-project/ray/blob/master/python/ray/data/dataset.py#L1872 .

**Comment by stephanie-wang** [comment]
> I think it's probably related to the `schema()` call inside `groupby` - https://github.com/ray-project/ray/blob/master/python/ray/data/dataset.py#L1872 .

Makes sense. I updated the issue description and priority to reflect this. Probably there is some nice way we can do limit pushdown for at least some all-to-all calls fairly easily, since we know the number of output rows for many of them (e.g., reparititon, sort).
